### PR TITLE
=act implement ByteString#takeRight(...)

### PR DESCRIPTION
**Pull request by monkey-mas**
Add implementation of takeRight(...).
### Benchmark

```
   --------------------------------- BASELINE ----------------------------------------------------------------------
   commit 0f2da7b26b5c4af35be87d2bd4a1a2392365df15
   [info] Benchmark                                                 Mode  Cnt          Score          Error  Units
   [info] ByteString_takeRight_Benchmark.bss_avg                   thrpt   40     486281.351 ±    14023.517  ops/s
   [info] ByteString_takeRight_Benchmark.bss_best                  thrpt   40     242096.480 ±     7029.451  ops/s
   [info] ByteString_takeRight_Benchmark.bss_greater_or_eq_to_len  thrpt   40  918217840.623 ± 12576483.168  ops/s
   [info] ByteString_takeRight_Benchmark.bss_negative              thrpt   40  895754845.562 ± 18965575.377  ops/s
   [info] ByteString_takeRight_Benchmark.bss_worst                 thrpt   40   11491250.938 ±   187949.479  ops/s

   --------------------------------- AFTER -------------------------------------------------------------------------

   ------ TODAY –––––––
   [info] Benchmark                                                 Mode  Cnt           Score          Error  Units
   [info] ByteString_takeRight_Benchmark.bss_avg                   thrpt   40      477028.892 ±     8846.994  ops/s
   [info] ByteString_takeRight_Benchmark.bss_best                  thrpt   40   106455228.296 ±   940597.222  ops/s
   [info] ByteString_takeRight_Benchmark.bss_greater_or_eq_to_len  thrpt   40   878842311.758 ±  9382511.848  ops/s
   [info] ByteString_takeRight_Benchmark.bss_negative              thrpt   40  1151544955.244 ± 21398125.610  ops/s
   [info] ByteString_takeRight_Benchmark.bss_worst                 thrpt   40      230406.055 ±     6075.782  ops/s
```

### CODE REVIEW

**Comment by jrudolph** [comment]
It seems the `avg` case and the `worst` case are actually slower than before? Or am I reading the stats the wrong way around?

**Comment by jrudolph** [comment]
OK TO TEST

**Comment by monkey-mas** [comment]
You're correct! :) My understanding is as follows (sorry for not having mentioned it):
I've replaced the implementation of `takeRight`, which was implemented using `slice(...)` == `drop(from).take(until - Math.max(0, from))`[1, 2],

- Before the replacement
  - `drop` and `take` are _optimized_ for dropping or taking elements from the left-most side,
  - In other words these methods are not _optimized_ for taking elements from the right-most side, which is not _good_ for `takeRight`
  - Thus `best` case has a _bad_ performance, on the other hand `worst` case has a _good_ performance
- After the replacement
  - `best` and `worst` cases: as I _optimized_ the implementation for taking elements from the right-most side, `best` case has a _good_ performance and `worst` case has a _bad_ performance like I expected.
  - `avg` case: Two things I can think of are:
    1. Like you said, now the performance is a little bit worse (almost 1.9% slower?)
    2. This might be observation error like the JMH `Error` value tells us.
 
Well, I _think_ the new implementation is what we can accept. But I might be wrong, let me know your thought please ;)

[1] https://github.com/akka/akka/pull/21564/files#diff-c8291b63bef2cf0ccb0123d3e15e3833L573
[2] https://github.com/monkey-mas/akka/blob/0afb7302b99ddf4d9109f99b08c50cee1dc16544/akka-actor/src/main/scala/akka/util/ByteString.scala#L269-L270

**Comment by jrudolph** [comment]
@monkey-mas thanks for the explanation. So, the underlying problem is that `ByteStrings` doesn't know the distribution of sizes in the array to be efficient in all cases. Instead, for degenerate cases it will have to traverse the whole (or a big part of the) list of subarrays to find the start and/or end component. I understand that there may be cases where, as a user, you know something about how a ByteString is composed to make use of the new `takeRight` to optimize accesses. However, I see the structure of a ByteString as internal knowledge and would try to prevent that users start to exploit this internal knowledge.

For complex ByteStrings with many parts, it would maybe make more sense e.g. to try if a binary search would be faster than what we currently have.

**Comment by monkey-mas** [comment]
@jrudolph Thanks for your comment! 

>  However, I see the structure of a ByteString as internal knowledge and would try to prevent that users start to exploit this internal knowledge.

Yes, as you wrote, users need to take full advantage of the internal implementation of `takeRight` to write an efficient code, which I first _thought_ is okay cos we have other methods like `take`, `drop` and so on and I expected users to make use of them. :)

> For complex ByteStrings with many parts, it would maybe make more sense e.g. to try if a binary search would be faster than what we currently have.

Do you have a specific data structure, algorithm or paper in your mind to give it a try? Your idea sounds like we might modify the current data structure(or algorithm) of `ByteString` so that we can have almost the same performance results between `best` and `worst` cases (in the end users really don't have to know and care about the internal knowledge when using `takeRight`). Do I have it correct? Or do I misunderstand your idea? ;)

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 1]
```diff
@@ -454,6 +465,29 @@ object ByteString {
...
+      val byteStringsSize = bytestrings.size
+      @tailrec def findSplit(fullTakes: Int, remainingToTake: Int): (Int, Int) = {
+        val bs = bytestrings(byteStringsSize - fullTakes - 1)
+        if (bs.length > remainingToTake) (fullTakes, remainingToTake)
+        else findSplit(fullTakes + 1, remainingToTake - bs.length)
+      }
+
+      val (fullTakes, remainingToTake) = findSplit(0, n)
+
+      if (fullTakes == 0)
```
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: If you put this logic inside of `findSplit` instead then there's no need to allocate the tuple

Comment (monkey-mas) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: Yes, you're right! Thanks for your feedback :)
So we can rewrite the logic like this?
```
private def takeRight0(n: Int): ByteString = {
  val byteStringsSize = bytestrings.size
  @tailrec def takeRightWithFullTakesAndRemaining (fullTakes: Int, remainingToTake: Int): ByteString = {
    val bs = bytestrings(byteStringsSize - fullTakes - 1)
    if (bs.length > remainingToTake) {
      if (fullTakes == 0)
        bytestrings(byteStringsSize - 1).takeRight(n)
      else if (remainingToTake == 0)
        new ByteStrings(bytestrings.takeRight(fullTakes), n)
      else
        new ByteStrings(bytestrings(byteStringsSize - fullTakes - 1).takeRight1(remainingToTake) +: bytestrings.takeRight(fullTakes), n)
    }
    else findSplit(fullTakes + 1, remainingToTake - bs.length)
  }

  takeRightWithFullTakesAndRemaining(0, n)
}
```

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 2]
```diff
@@ -454,6 +465,29 @@ object ByteString {
       else new ByteStrings(bytestrings.take(last) :+ bytestrings(last).take1(restToTake), n)
     }
 
+    override def takeRight(n: Int): ByteString =
+      if (n <= 0) ByteString.empty
+      else if (n >= length) this
+      else takeRight0(n)
```
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: Yes, precisely.

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 1 above)
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: Yes, something like that. :)

**Comment by patriknw** [comment]
my bad, I changed pr validation job params too early

**Comment by patriknw** [comment]
PLS BUILD

**Review by drewhk — APPROVED** [review]

**Comment by monkey-mas** [comment]
Can we merge this or should I close this? Let me know if somebody has a chance to look at it. I'll rebase my commits then if necessary :-)

**Comment by hepin1989** [comment]
@monkey-mas Could you please update it to be mergeable?

**Comment by hepin1989** [comment]
@patriknw @ktoso This PR need review to be merged.

**Comment by patriknw** [comment]
@monkey-mas I'm very sorry for how we have neglected this pull request. I think we are not fully convinced that the these changes improves performance for all cases and if it's worth changing (risking introducing something unwanted). We should have let you know much much earlier. If there is a real world use case where we see that this can help I'm open to revisiting this, but let's close it for now.

**Comment by monkey-mas** [comment]
sure, thanks for your review and help :)

**Pull request closed without merging**